### PR TITLE
Handle compiler panics when handling infinite lengths

### PR DESCRIPTION
**Pull request by bluebear94**
See #1918.

This currently fixes:

* NaN results when multiplying a zero `Abs` or `Em` unit with infinity, or multiplying an infinite unit with zero
* NaN results when adding or subtracting scalars such as `inf - inf` (results in 0)

### CODE REVIEW

**Comment by laurmaedje** [comment]
Having to deal with fallability across the whole layout code sounds like a huge pain. Rather than fixing individual panics across different layouters, I wonder whether we can find a general way to deal with this (perhaps by "normalizing" NaN to zero or something).

**Comment by bluebear94** [comment]
> perhaps by "normalizing" NaN to zero or something

That’s probably the simplest fix to the issues with arithmetic, though I’m concerned about this introducing surprising behavior (like `2pt + infpt - infpt != 2pt`).

**Comment by laurmaedje** [comment]
I think `2pt + infpt - infpt != 2pt` is okay. It just isn't well defined. That would also be a problem in pretty much every other programming language.

File: crates/typst/src/geom/scalar.rs [hunk 1]
```diff
@@ -6,6 +6,12 @@ use super::*;
 #[derive(Default, Copy, Clone)]
 pub struct Scalar(pub f64);
```
Comment (laurmaedje) [inline crates/typst/src/geom/scalar.rs:7]: Can this be made private to enforce the non-NaN invariant?

**Review by laurmaedje — COMMENTED** [review]
The panics with frames of infinite sizes probably still remain. Do you also want to address those in this PR or is it ready now?

**Comment by bluebear94** [comment]
> The panics with frames of infinite sizes probably still remain. Do you also want to address those in this PR or is it ready now?

I think these can be done in a future PR.

File: crates/typst/src/geom/scalar.rs [hunk 2]
```diff
@@ -4,7 +4,38 @@ use super::*;
...
 #[derive(Default, Copy, Clone)]
-pub struct Scalar(pub f64);
+pub struct Scalar(f64);
+
+// We have to detect NaNs this way since `f64::is_nan` isn’t const
+// on stable yet:
+// ([tracking issue](https://github.com/rust-lang/rust/issues/57241))
+#[allow(clippy::unusual_byte_groupings)]
+const fn is_nan_const(x: f64) -> bool {
+    let x_bits = unsafe { std::mem::transmute::<f64, u64>(x) };
```
Comment (laurmaedje) [inline crates/typst/src/geom/scalar.rs:14]: A safety comment that also explains that `to_bits` is not const fn would be nice.

**Comment by laurmaedje** [comment]
Thanks!

**Pull request merged**